### ISSUE TITLE
Add a new token to match the body of defcal

### ISSUE THREAD

**Issue by aspcompiler** [openqasm/openqasm#246]
### What is the expected enhancement?

TLDR: Match defcal body with a token so that defcal grammars can use their own lexer rules.

Details: Currently, the defcal body is parsed with a parser rule rather than a lexer rule. Since most of modern parsers separates the lexing and parsing phases,  defcal body is tokenized by the OpenQASM lexer rules. To allow other defcal grammar to implement their own lexer rules, it is desirable to match the defcal body as a single token and pass the resulting string to the parser of the defcal grammar.

Currently, the def grammar is written as:

```
calibrationDefinition
    : 'defcal' Identifier
    ( LPAREN calibrationArgumentList? RPAREN )? identifierList
    returnSignature? LBRACE .*? RBRACE  // for now, match anything inside body
    ;
```

`.*?` inside LBRACE and RBRACE will match any OpenQASM tokens non-greedily; this mean the body cannot contain RBRACE unless it is enclosed in a string or comment. White space and comments are skipped due to the OpenQASM lexer rule.

To preserve the body of defcal for other parsers, consider introducing a special lexer rule such as

```
ExternalGrammar
  : '{{' .*? '}}'
```

Then the defcal grammar would become:

```
calibrationDefinition
    : 'defcal' Identifier
    ( LPAREN calibrationArgumentList? RPAREN )? identifierList
    returnSignature? ExternalGrammar
    ;
```

**Comment by jakelishman** [comment]
A possible alternative to this is to use ANTLR's "island grammar" rules to swap the _lexing_ mode when the `defcal` token is encountered.  We can then have the lexer emit an entirely different set of tokens when it encounters braces; we could potentially use a lexeme made of fragments, which will allow all text with balanced braces to occur.  We'd probably want some way to embed a literal unbalanced brace in the text, if not only for completeness (handle `\{`, `\}` and `\\` as separate tokens?), and then restore them later during parse-tree simplification, before passing the body on to a separate lexer for `defcal` blocks.

Without the escape characters, this looks something approximately like:
```antlr
// switch to defcal mode when the defcal lexeme is encountered
defcal: 'defcal' -> mode(DEFCAL);

mode DEFCAL;
// ... tokens which can handle the defcal signature ...
// Consume a balanced brace block, and then return lexing back to QASM mode.
DefcalgrammarBody: DefcalgrammarBlock -> mode(DEFAULT_MODE);
fragment DefcalgrammarBlock: '{' ( ~('{' | '}') | DefcalgrammarBlock )* '}';
```

The advantage to this is that we don't need to introduce a special lexeme into the OpenQASM grammar to stop the lexer from incorrectly consuming non-QASM tokens.  This could also be used in `#pragma` blocks as well, which conceivably won't necessarily consist of valid OpenQASM tokens.  The obvious (imo minor) disadvantage is that it's marginally more complicated in the definition file, but I think it more accurately represents the underlying grammar.

**Comment by aspcompiler** [comment]
Island grammar is a good idea! According to [Antlr doc](https://github.com/antlr/antlr4/blob/master/doc/grammars.md), only lexer grammars can contain mode specifications. We will need to split our combined grammar into parser and lexer grammars. Is there any concerns before we do that?

**Comment by jakelishman** [comment]
Ah yes, I think you're right.  I hadn't noticed in the examples that it swapped from being the mixed-type parser/lexer in `lexer grammar ...;`.

I think the split should work cleanly, but I'm not dead certain.  The grammar file as it is doesn't use any of the abilities to insert host-language code into the lexing/parsing, so I don't think we can have accidentally created an unbreakable link between the two.  The output of `antlr -Dlanguage=Python3 qasm3.g4` generates two separate files for the lexer and parser, which seems promising.  I'm certainly no expert, though.

**Comment by jwoehr** [comment]
>The grammar file as it is doesn't use any of the abilities to insert host-language code into the lexing/parsing, so I don't think we can have accidentally created an unbreakable link between the two.

There has been some call for that, though. See #220

**Comment by jakelishman** [comment]
As I understand it, there's no problem in adding those features after a split. I was wondering if it were possible, had it been done while they were joined, that you could have tangled up the two states.

**Comment by aspcompiler** [comment]
I did an experiment to split the lexer and parser grammar. Unlike combined grammar, Antlr does not allow implicit token in parser grammar. For example, the following is not allowed in the parser grammar:

```
singleDesignatorType
    : 'int'
    | 'uint'
    | 'float'
    | 'angle'
    ;
```

We have to create tokens for INT, UNIT, FLOAT, ANGLE in the lexer grammar. Currently, we have many implicit tokens so we will be making a significant change to the grammar.

**Comment by stevenheidel** [comment]
This was approved by the TSC and we would welcome a PR that splits the lexer and parser grammar.

**Comment by jakelishman** [comment]
I'll do this in two parts: first I'll split the lexer and parser grammars in one PR, keeping names as close as possible, then I'll make a second PR that adds the defcal handling, since I think there'll be more to discuss on the latter, and I don't want to block the AST implementation while we thrash it out.
